### DISCUSSION TITLE
How to create Geometry in other thread?

### DISCUSSION THREAD

**Discussion by zcarey** [AvaloniaUI/Avalonia#15872] [Q&A]
I want to create some Geometry in background thread, because I need to deal with time-consuming operations, but it always throw an exception from VerifyAccess in AvaloniaObject :
`System.InvalidOperationException:“Call from invalid thread”`

Is there any way to create Geometry in another thread?

**Comment by Pratham95VP** [comment]
You can try something like this 
```
private async Task CreateGeometryAsync()
{
    var geometry = await Task.Run(() =>
    {       
        // Create the Geometry      
        return geometry;
    });

       await Dispatcher.UIThread.InvokeAsync(() =>
    {
        // Access the Geometry here
     
    });
}
``` 
UI objects like Geometry are not thread-safe and can only be accessed from the UI thread that's the reason why `System.InvalidOperationException:“Call from invalid thread”` this exception occurred.

**Reply by zcarey** [reply]
Thank you for your answer, but in this way it seems not to need to create another thread, I want to access Geometry in the background.

**Comment by damian-666** [comment]
im wondering if we should use Core2D this as a canonical  model,  but im a noob and i can just say it looks like its very well-architected, super clean, and been updated, but not sure if its using the newest apis, there's a couple bugs in it but thats its still a great sample of a full featured avalonia ide.   docking , tools, prop sheets, persistence, threading , drawings, viewmodel binding, immutables and threading .
i pulled build an run  it.  on devstudio newest on windows.   i think without issue.
, but i had to remove a reference to a daily build vault.

https://github.com/wieslawsoltes/Core2D

I could build and run , draw lots of shapes and save / load /undo/ redo , edit, adorn, them.

its normal to have to do access all visuals elements ( that are stateless views, wiht dependency properties, that listnen to a collision of points )  on the UI thread .    so the actual data can be loaded and changed, or processed,  in the workers, , then when a draw request comes it will take the last copy  , that the viewmodel is bound to , it seems like. 

so a lot of work with the  data  points and geometry can be done in worker threaded, then copied to immutable views that are bound to the viewmodels, so it can generate geom on multiple threads and draw the lastest copy .   This is like the ring buffer patter or producer consumer and doesn't require locking,     if you have lot of detall that might let you use SIMD or parallelism to organise winding of data points and sorting them around the centroid and tesselating .     a polygon with 199 grib sand really bog.   i would try to stress it it first.
generated 10k polygons and see if its ok.

BUG..  but if you undock , i cannot redock, im on 2 screens through.   an mess up the layout i have noy figure out how to reset it. 

please advise if there is a new sample , in Core2D there are command line switches i haven't explored... but i see scripting and i want to try to integrate KNI views into it.   and bepuphysics and aether physics.  And Avalonia edit for plugins.   this is  similar to whas stride wants also.

 as a game level editor ide or general test rig, strip it down a bit.   or use it as a template 
 
 please advise,  gurus.. i might be able to do a monogame sample. with multiple views  with this.. 
 
 @wieslawsoltes ?
   is it in good shape , should we try to build on this.. its not an "official sample " 
   but looks like a great way to see all what Avalonia has and how to organise it..

**Reply by zcarey** [reply]
This project seem COOL!!! I will study it! Thanks!
